Refactor run merging

mheinzel opened the pull request:
The refactoring I had planned, getting rid of the `IORef Entry` and enabling a few other optimisations.

I realised I had run the benchmarks in #203 using an older GHC version (9.2), so I also included that comparison here.

In terms of allocated bytes per entry:

| benchmark                      | GHC 9.2 | main  | refactor |
|--------------------------------|---------|-------|----------|
| word64-insert-x4               |     899 |   823 |      655 |
| word64-delete-x4               |     750 |   667 |      483 |
| word64-delete-x4-lastlevel     |     452 |   356 |      172 |
| word64-blob-x4                 |   31340 | 31198 |    31030 |
| word64-mupsert-x4              |     899 |   823 |      671 |
| word64-mupsert-collisions-x4   |     723 |   653 |      475 |
| word64-mix-x4                  |     849 |   771 |      603 |
| word64-mix-collisions-x4       |     652 |   581 |      413 |
| insert-large-keys-x4           |   34140 | 36680 |    36510 |
| insert-mixed-vals-x4           |   34380 | 36950 |    36780 |
| insert-page-x4                 |   51140 | 55030 |    54860 |
| insert-page-plus-byte-x4       |   98280 | 97840 |    97680 |
| insert-huge-vals-x4            |  194160 |193290 |   193130 |
| utxo-x4                        |    1473 |  1444 |     1268 |
| utxo-x4-lastlevel              |    1178 |  1130 |      954 |

Time is less stable, but there still seems to be a noticeable improvement (all in milliseconds per 50k entries, or 5k where indicated by `(*)`):

| benchmark                      | GHC 9.2 | main   | refactor |
|--------------------------------|---------|--------|----------|
| word64-insert-x4               |   13.83 |  13.76 |    13.01 |
| word64-delete-x4               |   12.73 |  13.03 |    12.71 |
| word64-delete-x4-lastlevel     |   5.992 |   5.58 |     4.87 |
| word64-blob-x4                 |  351.60 | 344.27 |   365.82 |
| word64-mupsert-x4              |   13.98 |  14.04 |    13.62 |
| word64-mupsert-collisions-x4   |   10.63 |  10.51 |     9.98 |
| word64-mix-x4                  |   14.88 |  15.22 |    14.72 |
| word64-mix-collisions-x4       |   11.18 |  11.12 |    10.77 |
| (*) insert-large-keys-x4       |   28.43 |  27.56 |    27.69 |
| (*) insert-mixed-vals-x4       |   27.21 |  27.40 |    26.98 |
| (*) insert-page-x4             |   39.70 |  39.87 |    39.83 |
| (*) insert-page-plus-byte-x4   |   72.79 |  71.67 |    73.45 |
| (*) insert-huge-vals-x4        |  133.00 | 131.01 |   133.86 |
| utxo-x4                        |   20.45 |  20.45 |    19.78 |
| utxo-x4-lastlevel              |   15.42 |  15.21 |    14.33 |


Full final results:

<details>

<summary>benchmark log</summary>

```
benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x2/merge
time                 11.71 ms   (11.68 ms .. 11.74 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 11.72 ms   (11.70 ms .. 11.76 ms)
std dev              80.15 μs   (44.77 μs .. 143.3 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.274e7    (3.274e7 .. 3.274e7)
  y                  79.130     (-1665.762 .. 1860.589)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x4/merge
time                 13.01 ms   (12.98 ms .. 13.04 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 13.05 ms   (13.04 ms .. 13.07 ms)
std dev              39.65 μs   (30.68 μs .. 55.32 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.277e7    (3.277e7 .. 3.277e7)
  y                  -71.319    (-1559.516 .. 1236.057)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x7/merge
time                 14.44 ms   (14.38 ms .. 14.51 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 14.33 ms   (14.31 ms .. 14.36 ms)
std dev              74.34 μs   (56.99 μs .. 104.3 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.281e7    (3.281e7 .. 3.281e7)
  y                  -83.184    (-1401.697 .. 1156.121)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x13/merge
time                 16.06 ms   (15.92 ms .. 16.18 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 15.97 ms   (15.94 ms .. 16.04 ms)
std dev              125.7 μs   (67.11 μs .. 204.8 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.294e7    (3.294e7 .. 3.294e7)
  y                  -3.959     (-11.234 .. 4.790e-7)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-delete-x4/merge
time                 12.71 ms   (12.67 ms .. 12.74 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 12.67 ms   (12.65 ms .. 12.70 ms)
std dev              76.34 μs   (57.33 μs .. 113.9 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.413e7    (2.413e7 .. 2.413e7)
  y                  -338.521   (-1520.617 .. 972.106)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-blob-x4/merge
time                 365.8 ms   (344.6 ms .. 409.0 ms)
                     0.998 R²   (0.997 R² .. 1.000 R²)
mean                 352.8 ms   (348.2 ms .. 360.5 ms)
std dev              7.320 ms   (259.8 μs .. 9.010 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              1.552e9    (1.552e9 .. 1.552e9)
  y                  -800.000   (-1600.000 .. 6.514e-6)
variance introduced by outliers: 19% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mupsert-x4/merge
time                 13.62 ms   (13.59 ms .. 13.67 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 13.67 ms   (13.64 ms .. 13.73 ms)
std dev              103.3 μs   (43.18 μs .. 189.6 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.357e7    (3.357e7 .. 3.357e7)
  y                  -68.148    (-2268.175 .. 2273.629)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mupsert-collisions-x4/merge
time                 9.975 ms   (9.955 ms .. 9.989 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 10.00 ms   (9.986 ms .. 10.05 ms)
std dev              73.55 μs   (27.33 μs .. 134.6 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.377e7    (2.377e7 .. 2.377e7)
  y                  77.593     (-2489.459 .. 2580.939)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mix-x4/merge
time                 14.72 ms   (14.69 ms .. 14.76 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 14.76 ms   (14.73 ms .. 14.86 ms)
std dev              112.0 μs   (48.87 μs .. 206.0 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.013e7    (3.013e7 .. 3.013e7)
  y                  -1172.462  (-2338.346 .. 2.152e-7)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mix-collisions-x4/merge
time                 10.77 ms   (10.75 ms .. 10.79 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 10.80 ms   (10.78 ms .. 10.89 ms)
std dev              92.31 μs   (36.90 μs .. 185.3 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.063e7    (2.063e7 .. 2.063e7)
  y                  -20.094    (-57.278 .. 2.221e-7)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-delete-x4-lastlevel/merge
time                 4.875 ms   (4.868 ms .. 4.881 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 4.879 ms   (4.876 ms .. 4.882 ms)
std dev              8.460 μs   (6.937 μs .. 10.77 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              8618808.736 (8618808.000 .. 8618809.669)
  y                  -25.408    (-56.114 .. 6.939e-8)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-large-keys-x4/merge
time                 27.70 ms   (27.62 ms .. 27.80 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 27.91 ms   (27.83 ms .. 28.01 ms)
std dev              194.6 μs   (153.5 μs .. 250.4 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              1.826e8    (1.826e8 .. 1.826e8)
  y                  -71.529    (-258.485 .. 73.587)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-mixed-vals-x4/merge
time                 26.98 ms   (26.89 ms .. 27.08 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 27.10 ms   (27.05 ms .. 27.17 ms)
std dev              123.2 μs   (93.18 μs .. 164.6 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              1.839e8    (1.839e8 .. 1.839e8)
  y                  100.601    (-8.217e-7 .. 243.108)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-page-x4/merge
time                 39.83 ms   (39.56 ms .. 40.15 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 39.84 ms   (39.68 ms .. 40.41 ms)
std dev              534.1 μs   (124.9 μs .. 987.4 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.743e8    (2.743e8 .. 2.743e8)
  y                  114.286    (-2.662e-6 .. 280.906)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-page-plus-byte-x4/merge
time                 73.45 ms   (73.21 ms .. 73.62 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 74.35 ms   (73.97 ms .. 74.96 ms)
std dev              840.9 μs   (274.9 μs .. 1.101 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.884e8    (4.884e8 .. 4.884e8)
  y                  471.467    (4.318e-6 .. 810.641)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-huge-vals-x4/merge
time                 133.9 ms   (133.5 ms .. 134.4 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 135.1 ms   (134.5 ms .. 136.8 ms)
std dev              1.494 ms   (144.6 μs .. 2.190 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              9.656e8    (9.656e8 .. 9.656e8)
  y                  137.143    (-9.562e-6 .. 278.400)
variance introduced by outliers: 12% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4/merge
time                 19.79 ms   (19.76 ms .. 19.82 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 19.84 ms   (19.80 ms .. 19.93 ms)
std dev              115.2 μs   (35.26 μs .. 228.1 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              6.341e7    (6.341e7 .. 6.341e7)
  y                  178.895    (-859.053 .. 1134.390)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4-uneven/merge
time                 19.78 ms   (19.73 ms .. 19.83 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 19.89 ms   (19.84 ms .. 20.01 ms)
std dev              173.5 μs   (59.68 μs .. 329.1 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              6.353e7    (6.353e7 .. 6.353e7)
  y                  84.876     (-912.941 .. 981.966)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4-lastlevel/merge
time                 14.33 ms   (14.30 ms .. 14.36 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 14.41 ms   (14.38 ms .. 14.55 ms)
std dev              142.3 μs   (41.13 μs .. 294.5 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.769e7    (4.769e7 .. 4.769e7)
  y                  -584.940   (-1742.689 .. 18.404)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4+1-min-skewed-lastlevel/merge
time                 14.75 ms   (14.72 ms .. 14.77 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 14.84 ms   (14.80 ms .. 14.92 ms)
std dev              120.9 μs   (52.12 μs .. 220.5 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.788e7    (4.788e7 .. 4.788e7)
  y                  12.026     (-178.448 .. 167.731)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4+1-max-skewed-lastlevel/merge
time                 14.50 ms   (14.26 ms .. 14.81 ms)
                     0.998 R²   (0.997 R² .. 1.000 R²)
mean                 14.23 ms   (14.14 ms .. 14.35 ms)
std dev              264.5 μs   (157.8 μs .. 378.9 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.795e7    (4.795e7 .. 4.795e7)
  y                  27.764     (-247.371 .. 306.049)
```

</details>


I still have some ideas for future optimisations around how `MutableHeap`, `ReadCtx` and `RunReader` play together, but that can wait.

It's also not quite ideal how blob inserts go from having a `BlobSpan` to `BlobRef` to `SerialisedBlob`. If we resolve blobs immediately, this saves 40 byte per entry on the reading side, but on the other hand we might read some blobs unnecessarily if their entries turn out to be obsolete. I think this will get easier with the changes to `RunReader` I have in mind.

jorisdral commented on src/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -199,13 +179,16 @@ close fs Merge {..} = do
       -- | Since there is always one reader outside of the heap, we need to
-      -- store it separately.
-    , readersLastRead :: !(IORef (RunReader fhandle, ReaderNumber))
+      -- store it separately. This also contains the next k\/op to yield, unless
+      -- all readers are drained, i.e. both:
+      -- 1. the reader it contains is empty
+      -- 2. the heap is empty
```
I think 2. implies 1., and the comment slightly confused me

mheinzel replied:
It's not the same! 1. talks about the reader inside the `ReadCtx` that is stored separately from the heap. I understand the confusion though and clarified the comment.

jorisdral replied:
Ah, since the reader in 1. is outside of the heap?

jorisdral commented on src/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -125,57 +72,90 @@ new fs mergeLevel mergeMappend targetPaths runs = do
+    (n, MergeInProgress) -> n >= requestedSteps
+    _                    -> True
+
+-- | Do at least a given number of steps of merging. Each step reads a single
+-- entry, then either resolves the previous entry with the new one or writes it
+-- out to the run being created.
+-- Returns the number of input entries read, which is guaranteed to be at least
+-- as many as requested (unless the merge is complete).
```
```suggestion
-- | Do at least a given number of steps of merging. Each step reads exactly one
-- entry, and outputs zero or one entries to the run under construction. 
-- Returns the number of input entries read, which is guaranteed to be at least
-- as many as requested (unless the merge is complete).
```

And maybe we should be even more explicit about why it might do more steps than requested

mheinzel commented on src/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -199,13 +179,16 @@ close fs Merge {..} = do
 --
--- Construct with 'newReaders', then keep calling 'getFromReaders'.
--- If aborting early, remember to call 'closeReaders'!
+-- Construct with 'readersNew', then keep calling 'readersPop'.
+-- If aborting early, remember to call 'readersClose'!
 data Readers fhandle = Readers {
```
So do you think a separate module would be better?

mheinzel replied:
> and how did you like `quickcheck-lockstep`?

There was quite a bit of ceremony. I'm sure all these parts have reasons for being there, to deal with issues arising in more complex examples, but for what I was doing (no variables, no opaque handle-like values that can't be compared etc.) it felt overkill. E.g. boilerplate like this:

```haskell
  observeReal _proxy = \case
      New {}          -> OEither . bimap OId OId
      PeekKey {}      -> OEither . bimap OId OId
      Pop {}          -> OEither . bimap OId (OTuple3 . trimap OId OId OId)
      DropWhileKey {} -> OEither . bimap OId (OTuple2 . bimap OId OId)
```

This is basically all `OId`, but if I just do that, the tests fail because it doesn't match `observeModel` (where I have to do all the bimap stuff for it to typecheck).

I could probably improve a few things, though. Let me push the tests to a separate branch, so you can have a look.

mheinzel replied:
https://github.com/IntersectMBO/lsm-tree/commit/b1e878ec7d0b190c317d637147dfa3b2fc1c32d9

mheinzel commented on src/Database/LSMTree/Internal/RunReader.hs:
```diff
@@ -85,21 +91,51 @@ close fs RunReader {..} = do
-    -- TODO(optimise): Sometimes, reading the overflow pages is not necessary.
-    -- We could just return the page index and offer a separate function to do
-    -- the disk I/O once needed.
-  | ReadLargeEntry
-      !SerialisedKey
-      !(Entry SerialisedValue (BlobRef (Run fhandle)))  -- ^ Value is only prefix.
-      !RawPage  -- ^ First page.
-      !Word32  -- ^ Length of the overflow in bytes.
+    RawEntryOverflow
+      -- | The value is just a prefix, with the remainder in the overflow pages.
+      !(Entry SerialisedValue (BlobRef (Run fhandle)))
+      -- | A page containing the single entry (or rather its prefix).
+      !RawPage
+      -- | Length of the overflow in bytes.
+      !Word32
+      -- | The overflow pages containing the suffix of the value.
+      --
+      -- TODO(optimise): Sometimes, reading the overflow pages is not necessary.
+      -- We could just return the page index and offer a separate function to do
+      -- the disk I/O once needed.
       ![RawOverflowPage]
```
Yes, it's a decent pattern, but overkill here. I'm even starting to think that the explicit invariant is overkill. We only have an invariant on the overflow case, which gets constructed in exactly one place.

So I think I'll just use a `mkEntryOverflow` smart constructor instead.

mheinzel commented on src/Database/LSMTree/Internal/RunReader.hs:
```diff
@@ -85,21 +91,51 @@ close fs RunReader {..} = do
   = Empty
-  | ReadSmallEntry !SerialisedKey !(Entry SerialisedValue (BlobRef (Run fhandle)))
-    -- | For large entries, the caller might be interested in various different
+  | ReadEntry !SerialisedKey !(RawEntry fhandle)
```
I'm not sure it's worth it here, though, since we go on to store the `RawEntry`, which means we'd have to re-box it.

mheinzel commented on test/Test/Database/LSMTree/Internal/Merge.hs:
```diff
@@ -68,20 +69,32 @@ prop_MergeDistributes fs level stepSize (fmap unTypedWriteBuffer -> wbs) = do
-      .&&. lhsBlobFile === rhsBlobFile
+           counterexample "runIndex"
+           (Run.runIndex lhs === Run.runIndex rhs)
+      .&&. counterexample "kops"
+           (lhsKOps === rhsKOps)
+      .&&. counterexample "kopsFile"
+           (lhsKOpsFile === rhsKOpsFile)
+      .&&. counterexample "blobFile"
+           (lhsBlobFile === rhsBlobFile)
```
Yes, but the error message for the first one is much more useful, so I ideally a bug would already be caught there.

mheinzel commented on src/Database/LSMTree/Internal/Merge.hs:
It very much depends on the workload, but if someone has many entries that are 2k-4k bytes large, it would happen a lot and when it happens, the improvement is significant.

I agree though, that it's probably not worth it. Should I remove the comment (or at rephrase it from a `TODO` to an observation)?

jorisdral replied:
Maybe explain your reasoning about 2k-4k byte entries in the comment. It wasn't obvious to me that we could benefit from this optimisation in the general case. We can keep the todo

Maybe the solution to this problem would instead be to increase the configured disk page size to something like 16k (even if the OS uses 4k disk pages), so that entries are more likely to end up on the same page

mheinzel commented on src/Database/LSMTree/Internal/Merge.hs:
Yes, but it is only relevant for large entries with blobs, which aren't common. I think it's a reasonably straightforward optimisation, so we might want to do it at some point, but it doesn't seem like a priority at the moment.

mheinzel commented on src/Database/LSMTree/Internal/RunReader.hs:
```diff
@@ -85,21 +91,51 @@ close fs RunReader {..} = do
-    -- | For large entries, the caller might be interested in various different
+  | ReadEntry !SerialisedKey !(RawEntry fhandle)
+
+data RawEntry fhandle =
```
Since this module is usually imported qualified, we might just have `Reader.Entry`. Changing it now!

mheinzel commented:
I addressed the feedback and rebased.

Note that the [changed target FPR on main](https://github.com/IntersectMBO/lsm-tree/commit/7fffe66774d7bd76615614d5d328b7e4468ee5fe#diff-fc1998087d84bb9fd8bed76dfac42ec31383bbf98e1c22386525f2334c477185) affects the allocation numbers above, but I will not update the table.

mheinzel commented:
I ended up rerunning the benchmarks, so now we have numbers with new Bloom filter FPR:


|                                | main   | refactor |
|--------------------------------|--------|----------|
| word64-insert-x4               |    921 |    753   |
| word64-delete-x4               |    764 |    580   |
| word64-blob-x4                 |  31295 |  31127   |
| word64-mupsert-x4              |    921 |    769   |
| word64-mupsert-collisions-x4   |    699 |    530   |
| word64-mix-x4                  |    868 |    700   |
| word64-mix-collisions-x4       |    628 |    460   |
| word64-delete-x4-lastlevel     |    358 |    174   |
| insert-large-keys-x4           |  36790 |  36620   |
| insert-mixed-vals-x4           |  37050 |  36880   |
| insert-page-x4                 |  55130 |  54970   |
| insert-page-plus-byte-x4       |  97940 |  97780   |
| insert-huge-vals-x4            | 193390 | 193230   |
| utxo-x4                        |   1541 |   1365   |
| utxo-x4-lastlevel              |   1179 |   1003   |

<details>

<summary>Full results</summary>

```
benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x2/merge
time                 14.51 ms   (14.35 ms .. 14.73 ms)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 14.54 ms   (14.43 ms .. 14.70 ms)
std dev              330.4 μs   (220.7 μs .. 474.7 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.760e7    (3.760e7 .. 3.760e7)
  y                  481.532    (-1406.739 .. 2538.567)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x4/merge
time                 16.17 ms   (15.89 ms .. 16.53 ms)
                     0.997 R²   (0.994 R² .. 0.999 R²)
mean                 16.11 ms   (15.92 ms .. 16.38 ms)
std dev              522.4 μs   (368.6 μs .. 724.7 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.763e7    (3.763e7 .. 3.763e7)
  y                  158.819    (-1481.987 .. 1562.817)
variance introduced by outliers: 12% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x7/merge
time                 17.10 ms   (16.89 ms .. 17.35 ms)
                     0.998 R²   (0.996 R² .. 1.000 R²)
mean                 17.49 ms   (17.27 ms .. 17.99 ms)
std dev              799.0 μs   (463.7 μs .. 1.353 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.767e7    (3.767e7 .. 3.767e7)
  y                  -290.277   (-1530.110 .. 1105.000)
variance introduced by outliers: 16% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-insert-x13/merge
time                 19.15 ms   (18.57 ms .. 19.69 ms)
                     0.998 R²   (0.996 R² .. 1.000 R²)
mean                 18.71 ms   (18.57 ms .. 18.97 ms)
std dev              458.1 μs   (263.8 μs .. 624.0 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.780e7    (3.780e7 .. 3.780e7)
  y                  -4.364     (-12.369 .. 4.544e-7)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-delete-x4/merge
time                 15.23 ms   (15.12 ms .. 15.36 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 15.21 ms   (15.16 ms .. 15.31 ms)
std dev              172.6 μs   (104.0 μs .. 280.2 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.899e7    (2.899e7 .. 2.899e7)
  y                  -238.827   (-1505.160 .. 1206.825)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-blob-x4/merge
time                 366.6 ms   (363.6 ms .. 371.3 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 372.9 ms   (369.8 ms .. 375.6 ms)
std dev              3.373 ms   (1.649 ms .. 4.443 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              1.556e9    (1.556e9 .. 1.556e9)
  y                  -800.000   (-1600.000 .. 7.664e-6)
variance introduced by outliers: 19% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mupsert-x4/merge
time                 16.24 ms   (15.99 ms .. 16.42 ms)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 16.48 ms   (16.32 ms .. 16.92 ms)
std dev              624.8 μs   (254.5 μs .. 1.134 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.843e7    (3.843e7 .. 3.843e7)
  y                  -153.069   (-2456.100 .. 2337.085)
variance introduced by outliers: 12% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mupsert-collisions-x4/merge
time                 11.35 ms   (11.25 ms .. 11.46 ms)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 11.30 ms   (11.25 ms .. 11.37 ms)
std dev              160.3 μs   (108.9 μs .. 228.9 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.651e7    (2.651e7 .. 2.651e7)
  y                  561.733    (-1710.863 .. 2597.219)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mix-x4/merge
time                 17.99 ms   (17.75 ms .. 18.20 ms)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 17.99 ms   (17.89 ms .. 18.11 ms)
std dev              262.9 μs   (198.6 μs .. 323.2 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              3.499e7    (3.499e7 .. 3.499e7)
  y                  -20.364    (-55.288 .. 4.309e-7)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-mix-collisions-x4/merge
time                 12.62 ms   (12.30 ms .. 12.97 ms)
                     0.997 R²   (0.995 R² .. 0.999 R²)
mean                 12.42 ms   (12.32 ms .. 12.58 ms)
std dev              313.7 μs   (203.1 μs .. 441.1 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.298e7    (2.298e7 .. 2.298e7)
  y                  -21.942    (-62.869 .. 3.174e-7)

benchmarking Bench.Database.LSMTree.Internal.Merge/word64-delete-x4-lastlevel/merge
time                 5.145 ms   (5.099 ms .. 5.217 ms)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 5.140 ms   (5.127 ms .. 5.163 ms)
std dev              56.17 μs   (38.77 μs .. 77.63 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              8677704.800 (8677704.000 .. 8677705.935)
  y                  -26.469    (-60.783 .. 8.037e-8)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-large-keys-x4/merge
time                 29.66 ms   (29.46 ms .. 29.93 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 29.73 ms   (29.59 ms .. 29.90 ms)
std dev              325.5 μs   (253.4 μs .. 460.5 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              1.831e8    (1.831e8 .. 1.831e8)
  y                  -1368.235  (-3191.871 .. -1.578e-6)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-mixed-vals-x4/merge
time                 28.97 ms   (28.56 ms .. 29.32 ms)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 29.07 ms   (28.86 ms .. 29.40 ms)
std dev              535.8 μs   (340.7 μs .. 732.3 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              1.844e8    (1.844e8 .. 1.844e8)
  y                  -323.176   (-1204.486 .. 861.003)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-page-x4/merge
time                 42.06 ms   (41.59 ms .. 42.50 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 42.34 ms   (42.08 ms .. 42.84 ms)
std dev              716.3 μs   (361.1 μs .. 982.9 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              2.748e8    (2.748e8 .. 2.748e8)
  y                  1834.901   (-2.553e-6 .. 3062.648)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-page-plus-byte-x4/merge
time                 76.89 ms   (76.34 ms .. 77.78 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 76.38 ms   (76.09 ms .. 76.93 ms)
std dev              678.4 μs   (312.5 μs .. 1.139 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              4.889e8    (4.889e8 .. 4.889e8)
  y                  505.600    (-2738.874 .. 3630.034)

benchmarking Bench.Database.LSMTree.Internal.Merge/insert-huge-vals-x4/merge
time                 139.9 ms   (137.9 ms .. 142.1 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 141.8 ms   (140.4 ms .. 143.1 ms)
std dev              1.946 ms   (1.455 ms .. 2.506 ms)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              9.662e8    (9.662e8 .. 9.662e8)
  y                  -457.143   (-917.073 .. 8.197e-6)
variance introduced by outliers: 12% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4/merge
time                 23.42 ms   (23.22 ms .. 23.62 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 23.29 ms   (23.20 ms .. 23.38 ms)
std dev              206.1 μs   (158.6 μs .. 275.2 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              6.827e7    (6.827e7 .. 6.827e7)
  y                  -759.747   (-2266.719 .. 1106.594)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4-uneven/merge
time                 23.30 ms   (23.16 ms .. 23.49 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 23.30 ms   (23.21 ms .. 23.41 ms)
std dev              226.7 μs   (168.4 μs .. 305.8 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              6.838e7    (6.838e7 .. 6.838e7)
  y                  334.653    (-785.331 .. 1288.224)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4-lastlevel/merge
time                 16.56 ms   (16.35 ms .. 16.75 ms)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 16.42 ms   (16.32 ms .. 16.54 ms)
std dev              274.1 μs   (210.2 μs .. 335.6 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              5.014e7    (5.014e7 .. 5.014e7)
  y                  51.731     (-3.771 .. 143.709)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4+1-min-skewed-lastlevel/merge
time                 16.81 ms   (16.68 ms .. 17.01 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 16.92 ms   (16.82 ms .. 17.06 ms)
std dev              280.4 μs   (185.7 μs .. 441.3 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              5.033e7    (5.033e7 .. 5.033e7)
  y                  9.138      (-273.062 .. 279.009)

benchmarking Bench.Database.LSMTree.Internal.Merge/utxo-x4+1-max-skewed-lastlevel/merge
time                 15.94 ms   (15.79 ms .. 16.14 ms)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 15.97 ms   (15.88 ms .. 16.08 ms)
std dev              248.6 μs   (152.6 μs .. 344.7 μs)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              5.042e7    (5.042e7 .. 5.042e7)
  y                  104.050    (-8.317 .. 205.755)
```

</details>
